### ISSUE TITLE
OCR0001: Creating Training data from Hand written Chuyik (8)

### ISSUE THREAD

**Issue by ta4tsering** [OpenPecha/ocr_handwriting_aligner#5]
## Description:
We have given two sets of worksheets, portrait mode and landscape mode. The annotators are instructed to write the Tibetan Cursive or Chuyik in the work sheet paper given. Now after reviewing and collecting the reviewed papers we need to do following steps:
   - Scan all the paper ( annotators work )
   - Write a script to crop all the reviewer accepted lines from each image
   - Name the line image as per its page or paper name and line number
   - and then create a csv for the line_name and its text.
   - Scaned pages [potrait](https://drive.google.com/drive/folders/1xUnktrh_ie0o-5vH_Grk8KN4rRl9uy0H), [landscape](https://drive.google.com/drive/folders/16FEH5m2Xu0TcX-JYNwc0iuuAaUaw1rLx)
   - [transcripts](https://drive.google.com/drive/folders/1ht0x3Ff86yDrsKKpQAi1I7prWnFH33Vt)
  
## Completion Criteria:
- accepted line images and a csv containing the line image and text

## Implementation Workflow:

![Image](https://github.com/OpenPecha/create_ocr_data/assets/52460417/781bef01-22d7-419e-9e5a-a001aadf5d42)

## Work items:

- [x] script to split pdf to jpg
- [x] check if the page ordering is consistent
- [x] find bounding positions for potrait
- [x] crop the images with positions with mapping
- [x] classifying line image (good or bad / black dot or no black dot)
- [x] map good line image with transcript
- [x] find bounding positions for landscape
- [x] modify the pipeline and run
- [x] find better method to find image margin
- [x] upload files to s3 bucket

**Comment by tenzin3** [comment]
For the second work item "**check if the page ordering is consistent**", i split the pdf **P000010_v001_00001 - 00500.pdf** into images and 500 images were in the output. So the number count matches, as for the number in each pdf page, i manually scrolled for a minute to check if it ordering is in proper manner. and so far is does.

**Comment by ta4tsering** [comment]
yeah it sounds like they are in order, they were told to put them in order. you can go ahead and start with it. and as for getting the line images bounding box, maybe you can use transkribus to manually annotate one or two images and then use the bounding box output in xml from the transkribus to use for cropping out the line images. if you want my help with line annotation in the transkribus let me know. [trranskribus download](https://readcoop.eu/transkribus/download/). create an account and download the application and open it in your computer. we can meet tomorrow to annotate the lines.

**Comment by tenzin3** [comment]
Thank you @ta4tsering , would look into that.

**Comment by tenzin3** [comment]
To classifying the line image quality, 

### Good Image
![Image](https://github.com/OpenPecha/create_ocr_data/assets/52460417/6a803e6d-bebf-4811-b784-5e6d833a4fd8)
### Bad Image
![Image](https://github.com/OpenPecha/create_ocr_data/assets/52460417/6a16e27a-b0a8-4882-b4af-439774b6d14b)



### Black to white ration for 3500 image (Potrait)
![Image](https://github.com/OpenPecha/create_ocr_data/assets/52460417/e8d6af8d-5796-4c4d-9346-cb5e0c27b43d)

Setting threshold of 0.125 such that
    >0.125 -> Good image, 
    <=0.125 ->Bad image, 
is great result



![Image](https://github.com/OpenPecha/create_ocr_data/assets/52460417/8f7e6934-3be7-4157-8455-dcd7c84ef6db)


### Problem 


![Image](https://github.com/OpenPecha/create_ocr_data/assets/52460417/4d5921d8-895e-4cb7-b5cf-305908b85f77)

### Reason to the problem
The reason is due to the some difference in gaps in the starting of the scanned pages, ~(8 out of 500 images) so 7*8 = 56 images
![Image](https://github.com/OpenPecha/create_ocr_data/assets/52460417/2d92475e-b693-46c4-8861-6130cfbcdd0d)

### Solution
Bigger bounding position for the line image label

**Comment by tenzin3** [comment]
To solve the above problem, rather than creating a bounding position to get the circle at center of the image, a new bounding position where the circle is vertically slightly lower than center. This way the circle may be be different position but it will be present in the image without the border line. (size of the bounding position has also been changed in the process leading to different threshold) 


### image example 1
![Image](https://github.com/OpenPecha/ocr_handwriting_aligner/assets/52460417/7f5394a7-a689-4159-8754-28ac9a802a05)
![Image](https://github.com/OpenPecha/ocr_handwriting_aligner/assets/52460417/47b30d65-78ab-49fb-ba64-60346cb8a58f)

### image example 2


![Image](https://github.com/OpenPecha/ocr_handwriting_aligner/assets/52460417/70125eda-221a-4fe1-9fa3-715964f7d84a)
![Image](https://github.com/OpenPecha/ocr_handwriting_aligner/assets/52460417/599884b8-0b14-4f95-b888-33d3031f2be8)



### Updated Displot 
Threshold set to 0.04

![Image](https://github.com/OpenPecha/ocr_handwriting_aligner/assets/52460417/100942a9-2d2a-424c-912b-a14d63a96d8a)

**Comment by tenzin3** [comment]
### Exception case
 **P000010_v001_00001 - 00500_to_image_000192.jpg** , more adjustment would let to more false negative.(circle would go outside the image for other image)
![Image](https://github.com/OpenPecha/ocr_handwriting_aligner/assets/52460417/3ecad671-fe5b-4a49-9139-b2250a3b5954)


### False positive


![Image](https://github.com/OpenPecha/ocr_handwriting_aligner/assets/52460417/afe27662-0595-4041-bbdc-e5b9fe1e1fc2)
![Image](https://github.com/OpenPecha/ocr_handwriting_aligner/assets/52460417/ad0b1a8b-1e58-432d-96b1-13dd6d092142)


### False negative


![Image](https://github.com/OpenPecha/ocr_handwriting_aligner/assets/52460417/50401c2a-4448-4d31-9351-1ed328c06c52)

**Comment by tenzin3** [comment]
![Image](https://github.com/OpenPecha/ocr_handwriting_aligner/assets/52460417/d564a4cf-74bb-4234-9282-741433858dba)

**Comment by tenzin3** [comment]
### Included image cleaning (Morphological Closing) in OpenCV and improved in getting proper image margin.

Updated coordinate for potrait and landscape image label coordinates. 

The black and white displot for potrait remains same. 
For landscape the displot for black to white ratio is below


![Image](https://github.com/OpenPecha/ocr_handwriting_aligner/assets/52460417/eb307996-dc9d-4d74-8181-b6f4b665a9ee)

**Comment by tenzin3** [comment]
Done

Landscape====>
P00013_v001_(1-353)
P00016_v001_(501-593)

Portrait======>
P00010_v001 (1-500,	500-1000,1000-1456)

**Comment by tenzin3** [comment]
Doubt:> 

Portrait======>
P00011_v001
(1-500 (some of the pages doesn't have label eg page no 15),	500-1000,		1000-1340)
P00014_v001(1-163)  [Some page don’t have labels, eg page no 1]

P00015_v001(50-250) [Some page don’t have labels, eg page 7 ]

Some of the page in the above volume doesnt have proper labels.
## Example 
![Image](https://github.com/OpenPecha/ocr_handwriting_aligner/assets/52460417/f46da837-8332-40d1-8b83-d49b851c26e1)
